### PR TITLE
Add new deep learning functionality

### PR DESCRIPTION
**Pull request by ncguilbeault**
This PR adds new functionality into the `Bonsai.ML.Torch` package to allow creation of deep learning modules directly in Bonsai. Operators were added to the `Bonsai.ML.Torch.NeuralNets` namespace to support a wide range of deep learning functionality, including allowing the direct creation of torch modules (convolutional layers, pooling layers, padding layers, recurrent neural networks, transformers, etc), and existing operators were refactored to support this new functionality in addition to previous functionality.

The PR in https://github.com/bonsai-rx/machinelearning/pull/77 should be merged first before this one.

### CODE REVIEW

File: src/Bonsai.ML.Torch/NeuralNets/ActivationFunction/Elu.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.ActivationFunction;
+
+/// <summary>
+/// Represents an operator that creates an exponential linear unit (ELU) activation function.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.ELU.html"/> for more information.
+/// </remarks>
+[Description("Creates an exponential linear unit (ELU) activation function.")]
+[DisplayName("ELU")]
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/ActivationFunction/Elu.cs]: Same comment as above.

File: src/Bonsai.ML.Torch/NeuralNets/ActivationFunction/Gelu.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.ActivationFunction;
+
+/// <summary>
+/// Represents an operator that creates a gaussian error linear unit (GELU) activation function.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.GELU.html"/> for more information.
+/// </remarks>
+[Description("Creates a gaussian error linear unit (GELU) activation function.")]
+[DisplayName("GELU")]
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/ActivationFunction/Gelu.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/Convolution/Conv2d.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.Convolution;
+
+/// <summary>
+/// Represents an operator that creates a 2D convolution module.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html"/> for more information.
+/// </remarks>
+[Description("Creates a 2D convolution module.")]
+public class Conv2d
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Convolution/Conv2d.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/Convolution/Fold.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+/// </remarks>
+[Description("Creates a fold module.")]
+public class Fold
+{
+    /// <summary>
+    /// The shape of the spatial dimensions of the output tensor.
+    /// </summary>
+    [Description("The shape of the spatial dimensions of the output tensor.")]
+    [TypeConverter(typeof(ValueTupleConverter<long, long>))]
+    public (long, long) OutputSize { get; set; }
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Convolution/Fold.cs]: Do tuples work well in the property grid and XML? I never tried, so curious. I wonder though if we should just use one of the size structures for this. I could swear we had something that would work here, and would be great to have some consistency.

File: src/Bonsai.ML.Torch/NeuralNets/Loss/BinaryCrossEntropyWithLogits.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+/// <summary>
+/// Represents an operator that creates a binary cross entropy with logits (BCEWithLogits) loss module.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html"/> for more information.
+/// </remarks>
+[Description("Creates a binary cross entropy with logits (BCEWithLogits) loss module.")]
+[TypeConverter(typeof(TensorOperatorConverter))]
+[DisplayName("BCEWithLogits")]
+public class BinaryCrossEntropyWithLogits : IScalarTypeProvider
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Loss/BinaryCrossEntropyWithLogits.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/Loss/MeanSquaredError.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.Loss;
+
+/// <summary>
+/// Represents an operator that creates a mean squared error (MSE) loss module.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.MSELoss.html"/> for more information.
+/// </remarks>
+[Description("Creates a mean squared error (MSE) loss module.")]
+[DisplayName("MSE")]
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Loss/MeanSquaredError.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/Normalization/InstanceNorm2d.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.Normalization;
+
+/// <summary>
+/// Represents an operator that creates a 2d instance normalization module.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.InstanceNorm2d.html"/> for more information.
+/// </remarks>
+[Description("Creates a 2d instance normalization module.")]
+public class InstanceNorm2d
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Normalization/InstanceNorm2d.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/Pooling/AdaptiveMaxPool2d.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.Pooling;
+
+/// <summary>
+/// Represents an operator that creates a 2D adaptive max pooling module.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveMaxPool2d.html"/> for more information.
+/// </remarks>
+[Description("Creates a 2D adaptive max pooling module.")]
+public class AdaptiveMaxPool2d
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Pooling/AdaptiveMaxPool2d.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/Pooling/FractionalMaxPool3d.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+namespace Bonsai.ML.Torch.NeuralNets.Pooling;
+
+/// <summary>
+/// Represents an operator that creates a 3D fractional max pooling module.
+/// </summary>
+/// <remarks>
+/// See <see href="https://pytorch.org/docs/stable/generated/torch.nn.FractionalMaxPool3d.html"/> for more information.
+/// </remarks>
+[Description("Creates a 3D fractional max pooling module.")]
+public class FractionalMaxPool3d
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/Pooling/FractionalMaxPool3d.cs]: Same as above.

File: src/Bonsai.ML.Torch/NeuralNets/ShuffleModuleBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+using System.Xml.Serialization;
+
+namespace Bonsai.ML.Torch.NeuralNets;
+
+/// <summary>
+/// Represents an operator that creates a module for shuffling input.
+/// </summary>
+[XmlInclude(typeof(Shuffle.ChannelShuffle))]
+[DefaultProperty(nameof(ShuffleModule))]
+[Combinator]
```
Comment (glopesdev) [inline src/Bonsai.ML.Torch/NeuralNets/ShuffleModuleBuilder.cs]: Same as above.

**Review by glopesdev — COMMENTED** [review]
Massive effort, thanks for putting this all together! Lots of comments below mainly on naming and organization for us to think about.

File: src/Bonsai.ML.Torch/NeuralNets/Convolution/Fold.cs (hunk 4 above)
Comment (ncguilbeault) [reply src/Bonsai.ML.Torch/NeuralNets/Convolution/Fold.cs:23]: Perhaps the [NumericRecordConverter](https://github.com/bonsai-rx/bonsai/blob/87fa801597d22b5ca0246f2e936ceda287dc8cc8/Bonsai.Core/NumericRecordConverter.cs#L13) class would work better here
